File: 3-LongestSubstringWithoutRepeatingCharacters.java
import java.util.Map;
import java.util.HashMap;

class Solution {
public static int lengthOfLongestSubstring(String s) {
          int res = 0;
        int start = 0;
        int end = 0;
        Map<Character, Integer> charMap = new HashMap<>();

        while (end < s.length()) {
            char c = s.charAt(end);
            if (charMap.containsKey(c)) {
                start = Math.max(start, charMap.get(c));
            }
            res = Math.max(res, end - start + 1);
            charMap.put(c, end + 1);
            end++;
        }
        return res;
    }
}

File: 3-LongestSubstringWithoutRepeatingCharacters.py
class Solution:
  def lengthOfLongestSubstring(self, s: str) -> int:
    left = right = 0
    maxLen = 0
  
    # print(f"len {len(s)}")
    # print(f"first {s[0]}")
    # print(f"right: {right}")
    curr = set()
    while right < len(s):
      c = s[right]
      
      if c not in curr:
        curr.add(c)
        right += 1
      else:
        maxLen = max(maxLen, len(curr))
        curr.clear()
        left += 1
        right = left
      # print(f"curr: {curr}")
    return max(maxLen, len(curr))

      # curr = 1
      # if s[right] not in s[left:right]:
      #   curr += 1
      # else:
      #   left = right
      #   maxLen = max(curr, maxLen)
      # right += 1
    